ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/AbstractBlockParameterOrBlockHashMethod.java

```diff
@@ -0,0 +1,117 @@
+import org.hyperledger.besu.ethereum.api.jsonrpc.internal.response.JsonRpcSuccessResponse;
+import org.hyperledger.besu.ethereum.api.query.BlockchainQueries;
+import org.hyperledger.besu.ethereum.core.Hash;
+
+import java.util.function.Supplier;
+
+import com.google.common.base.Suppliers;
+
+public abstract class AbstractBlockParameterOrBlockHashMethod implements JsonRpcMethod {
+  private static final String INVALID_BLOCK_NUMBER_PARAMETER = "Invalid Block Number Parameter.";
+
+  protected final Supplier<BlockchainQueries> blockchainQueries;
+
+  protected AbstractBlockParameterOrBlockHashMethod(final BlockchainQueries blockchainQueries) {
+    this(Suppliers.ofInstance(blockchainQueries));
+  }
+
+  protected AbstractBlockParameterOrBlockHashMethod(
+      final Supplier<BlockchainQueries> blockchainQueries) {
+    this.blockchainQueries = blockchainQueries;
+  }
+
+  protected abstract BlockParameterOrBlockHash blockParameterOrBlockHash(
+      JsonRpcRequestContext request);
+
+  protected abstract Object resultByBlockNumber(JsonRpcRequestContext request, long blockNumber);
+
+  protected BlockchainQueries getBlockchainQueries() {
+    return blockchainQueries.get();
+  }
+
+  protected Object pendingResult(final JsonRpcRequestContext request) {
+    // TODO: Update once we mine and better understand pending semantics.
+    // For now act like we are not mining and just return latest.
+    return latestResult(request);
+  }
+
+  protected Object latestResult(final JsonRpcRequestContext request) {
+    return resultByBlockNumber(request, blockchainQueries.get().headBlockNumber());
+  }
+
+  protected Object handleParamTypes(final JsonRpcRequestContext requestContext) {
+    final BlockParameterOrBlockHash blockParameterOrBlockHash =
+        blockParameterOrBlockHash(requestContext);
+
+    final Object result;
+    if (blockParameterOrBlockHash.isLatest()) {
+      result = latestResult(requestContext);
+    } else if (blockParameterOrBlockHash.isPending()) {
+      result = pendingResult(requestContext);
+    } else if (blockParameterOrBlockHash.isNumeric() || blockParameterOrBlockHash.isEarliest()) {
+      result =
+          resultByBlockNumber(
+              requestContext,
+              blockParameterOrBlockHash
+                  .getNumber()
+                  .orElseThrow(() -> new IllegalStateException(INVALID_BLOCK_NUMBER_PARAMETER)));
+    } else {
+      Hash blockHash =
+          blockParameterOrBlockHash
+              .getHash()
+              .orElseThrow(() -> new IllegalStateException(INVALID_BLOCK_NUMBER_PARAMETER));
+      if (Boolean.TRUE.equals(blockParameterOrBlockHash.getRequireCanonical())
+          && !blockchainQueries.get().blockIsOnCanonicalChain(blockHash)) {
+        return new JsonRpcErrorResponse(
+            requestContext.getRequest().getId(), JsonRpcError.JSON_RPC_NOT_CANONICAL_ERROR);
+      }
+
+      result =
+          resultByBlockNumber(
```

Comment:
If a hash on an orphan chain is passed in the net result of hash to block number to sub call is that we will do the call on the wrong block.